packages/local_auth/local_auth_android/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java

```diff
@@ -147,79 +146,78 @@ private void authenticate(MethodCall call, final Result result) {
-              lifecycle, (FragmentActivity) activity, call, completionHandler, true);
-      authHelper.authenticate();
-      return;
+    // Unable to authenticate
+    int errorCode =
+        isBiometricOnly
+            ? biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK)
+            : biometricManager.canAuthenticate(
+                BiometricManager.Authenticators.BIOMETRIC_WEAK
+                    | BiometricManager.Authenticators.DEVICE_CREDENTIAL);
+
+    switch (errorCode) {
+      case BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE:
+        completionHandler.onError("NoHardware", "No biometric hardware found");
+        break;
+      case BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED:
+        completionHandler.onError(
+            "NotEnrolled", "No requested authenticators enrolled on this device");
+        break;
+      case BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE:
+        completionHandler.onError(
+            "NotAvailable", "No requested authenticators supported or enabled");
+        break;
+      case BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED:
+        completionHandler.onError("NotSupported", "This device requires a security update");
+        break;
+      default:
+        result.error(
+            "NotSupported", "This device does not support required security features", null);
+        break;
     }
 
-    // API 23 - 28 with fingerprint
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && fingerprintManager != null) {
-      if (fingerprintManager.hasEnrolledFingerprints()) {
-        authHelper =
-            new AuthenticationHelper(
-                lifecycle, (FragmentActivity) activity, call, completionHandler, false);
-        authHelper.authenticate();
-        return;
+    return;
+  }
+
+  @VisibleForTesting
+  public AuthCompletionHandler createAuthCompletionHandler(final Result result) {
+    return new AuthCompletionHandler() {
+      @Override
+      public void onSuccess() {
+        authenticateSuccess(result);
       }
-    }
 
-    // API 23 or higher with device credentials
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
-        && keyguardManager != null
-        && keyguardManager.isDeviceSecure()) {
-      String title = call.argument("signInTitle");
-      String reason = call.argument("localizedReason");
-      Intent authIntent = keyguardManager.createConfirmDeviceCredentialIntent(title, reason);
-
-      // save result for async response
-      lockRequestResult = result;
-      activity.startActivityForResult(authIntent, LOCK_REQUEST_CODE);
-      return;
-    }
+      @Override
+      public void onFailure() {
+        authenticateFail(result);
+      }
 
-    // Unable to authenticate
-    result.error("NotSupported", "This device does not support required security features", null);
+      @Override
+      public void onError(String code, String error) {
+        if (authInProgress.compareAndSet(true, false)) {
```

Comment:
I see that this is pre-existing, but do you know why only error cases set this back to false? Is there another call elsewhere for success/failure?

Reply:
Yeah, the other callbacks call helper methods that set it back to false, as well!